### ISSUE TITLE
CudaGraphs Support

### ISSUE THREAD

**Issue by crtrott** [kokkos/kokkos#2132]
This is an issue to discuss CUDA graphs support.

**Comment by crtrott** [comment]
Test Code
```c++
#include<Kokkos_Core.hpp>
#include<cmath>

template<class result_space_type>
void run_test(int N, int R) {
    Kokkos::View<int*> a("A",N);

    cudaStream_t stream1;
    cudaGraph_t graph;
    int64_t result_h = 0.0;

    // Create a stream
    cudaStreamCreate(&stream1);
    Kokkos::Cuda cuda1(stream1);


    Kokkos::parallel_for("Init",N,KOKKOS_LAMBDA (const int i) {
      a(i) = i;
    });

    Kokkos::View<int64_t, result_space_type> result("Result");

    Kokkos::fence();

      Kokkos::parallel_for("Add-5",Kokkos::RangePolicy<>(cuda1,0,N),KOKKOS_LAMBDA (const int i) {
        a(i) += 5;
      });

      Kokkos::parallel_for("Sub-3",Kokkos::RangePolicy<>(cuda1,0,N),KOKKOS_LAMBDA (const int i) {
       a(i) -= 3;
      });

      Kokkos::parallel_reduce("reduce",Kokkos::RangePolicy<>(cuda1,0,N),KOKKOS_LAMBDA (const int i, int64_t& lsum) {
       lsum += a(i);
      },result);
      cudaStreamSynchronize(stream1);

    Kokkos::parallel_for("Init",N,KOKKOS_LAMBDA (const int i) {
      a(i) = i;
    });
    Kokkos::fence();

    Kokkos::Timer timer;
    for(int r=0; r<R; r++) {
      Kokkos::parallel_for("Add-5",Kokkos::RangePolicy<>(cuda1,0,N),KOKKOS_LAMBDA (const int i) {
        a(i) += 5;
      });

      Kokkos::parallel_for("Sub-3",Kokkos::RangePolicy<>(cuda1,0,N),KOKKOS_LAMBDA (const int i) {
       a(i) -= 3;
      });

      Kokkos::parallel_reduce("reduce",Kokkos::RangePolicy<>(cuda1,0,N),KOKKOS_LAMBDA (const int i, int64_t& lsum) {
       lsum += a(i);
      },result);
      int64_t s_result;
      if(std::is_same<result_space_type,Kokkos::CudaSpace>::value) {
        Kokkos::deep_copy(cuda1,s_result,result);
        cudaStreamSynchronize(stream1);
      } else {
        cudaStreamSynchronize(stream1);
        s_result = result();
      }
      printf("Result %i %li %li\n",r,s_result,int64_t(N)*int64_t(N-1)/2+int64_t(N)*2*(r+1));
    }
    double time = timer.seconds();
    printf("NoGraphTime<%s>: %e\n",result_space_type::name(),time);

    Kokkos::parallel_for("Init",N,KOKKOS_LAMBDA (const int i) {
      a(i) = i;
    });
    Kokkos::fence();
    // Start capturing stream
    cudaStreamBeginCapture(stream1, cudaStreamCaptureModeGlobal);
    {
      Kokkos::parallel_for("Add-5",Kokkos::RangePolicy<>(cuda1,0,N),KOKKOS_LAMBDA (const int i) {
        a(i) += 5;
      });

      Kokkos::parallel_for("Sub-3",Kokkos::RangePolicy<>(cuda1,0,N),KOKKOS_LAMBDA (const int i) {
       a(i) -= 3;
      });

      Kokkos::parallel_reduce("reduce",Kokkos::RangePolicy<>(cuda1,0,N),KOKKOS_LAMBDA (const int i, int64_t& lsum) {
       lsum += a(i);
      },result);
    }
    cudaStreamEndCapture(stream1, &graph);

    // Create graphExec
    cudaGraphExec_t graphExec;
    cudaGraphInstantiate(&graphExec, graph, NULL, NULL, 0);

    // Run graph
    Kokkos::fence();
    timer.reset();
    for(int r=0; r<R; r++) {
      cudaGraphLaunch(graphExec, stream1);
      int64_t s_result;
      if(std::is_same<result_space_type,Kokkos::CudaSpace>::value) {
        Kokkos::deep_copy(cuda1,s_result,result);
        cudaStreamSynchronize(stream1);
      } else {
        cudaStreamSynchronize(stream1);
        s_result = result();
      }
      printf("Result %i %li %li\n",r,s_result,int64_t(N)*int64_t(N-1)/2+int64_t(N)*2*(r+1));
    }
    double time2 = timer.seconds();
    printf("GraphTime<%s>: %e\n",result_space_type::name(),time2);
    Kokkos::fence();

    // Check whether its ok
    Kokkos::parallel_for("Check",N,KOKKOS_LAMBDA (const int i) {
      if(a(i)!=i+R*2) printf("Error: %i %i %i\n",i,a(i),i+R*2);
    });
}

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv);
  {
    int N = argc>1?atoi(argv[1]):1000;
    int R = argc>2?atoi(argv[2]):10;
    int T = argc>3?atoi(argv[3]):0;

    if(T==0)
       run_test<Kokkos::CudaHostPinnedSpace>(N,R);
    if(T==1)
       run_test<Kokkos::CudaSpace>(N,R);
    if(T==2)
       run_test<Kokkos::HostSpace>(N,R);
  }
  Kokkos::finalize();
}
```
```

**Comment by PhilMiller** [comment]
It looks like the `WorkGraph` stuff may have addressed this, and was indeed present in the 3.3 release as the Milestone indicates, from PR #3369

**Comment by PhilMiller** [comment]
Ugh, I'm trying to re-add Daisy to give attribution, but Github isn't letting me.
